### PR TITLE
[KVStore] add init_data() on client

### PR DESCRIPTION
**Pull request by aksnzhy**
## Description

Server side:

```
    my_server = KVServer(server_id=0, server_namebook=server_namebook, num_client=1)
    my_server.set_global2local(name='data_0', global2local=g2l_0)
    my_server.set_global2local(name='data_1', global2local=g2l_1)
    my_server.set_partition_book(name='data_0', partition_book=partition_0)
    my_server.set_partition_book(name='data_1', partition_book=partition_1)
    my_server.init_data(name='data_0', data_tensor=data_0)
    my_server.init_data(name='data_1', data_tensor=data_1)

    my_server.start()
```

client side:

```
    my_client = KVClient(server_namebook=server_namebook)
    my_client.connect()
    # data_2 has the same partition_book and g2l with data_0 by setting target_name = 'data_0'
    my_client.init_data(name='data_2', shape=my_shape, dtype=F.float32, target_name='data_0')
```

Then the client can access `data_0`, `data_1`, and `data_2` . Note that,` my_client.init_data()` should be invoked by all the clients, and all the clients will be blocked until the server and client finish initialization. Then, the client can get their own data partition.

For now, all the new data will be initialized to zeros.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/contrib/dis_kvstore.py [hunk 1]
```diff
@@ -473,6 +473,29 @@ def start(self):
                     data=res_tensor,
                     c_ptr=None)
                 _send_kv_msg(self._sender, back_msg, msg.rank)
+            # Init new data
+            elif msg.type == KVMsgType.NEW_DATA:
+                assert msg.rank == 0
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py:478]: only one process sends the message?

File: python/dgl/contrib/dis_kvstore.py [hunk 2]
```diff
@@ -757,6 +802,71 @@ def connect(self):
...
+        Parameters
+        ----------
+        name : str
+            data name
+        shape : list of int
+            data shape
+        dtype : dtype
+            data type
+        target_name : str
+            target name is used to find existing partition_book and g2l mapping.
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py:822]: what is this? we had a discussion that we should group tensors into ID spaces.

File: python/dgl/contrib/dis_kvstore.py [hunk 3]
```diff
@@ -473,6 +473,29 @@ def start(self):
...
+            # Init new data
+            elif msg.type == KVMsgType.NEW_DATA:
+                assert msg.rank == 0
+                data_str, target_name = msg.name.split('|')
+                data_name, data_type = self._deserialize_shared_tensor(data_str)
+                dtype = F.data_type_dict[data_type]
+                data_shape = F.asnumpy(msg.id).tolist()
+                if self._server_id % self._group_count == 0: # master server
+                    data_tensor = F.zeros(data_shape, dtype, F.cpu())
+                    self.init_data(name=data_name, data_tensor=data_tensor)
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py:485]: `init_data` will copy the tensor to the shared memory, right?

File: python/dgl/contrib/dis_kvstore.py [hunk 4]
```diff
@@ -473,6 +473,29 @@ def start(self):
                     data=res_tensor,
                     c_ptr=None)
                 _send_kv_msg(self._sender, back_msg, msg.rank)
+            # Init new data
+            elif msg.type == KVMsgType.NEW_DATA:
+                assert msg.rank == 0
+                data_str, target_name = msg.name.split('|')
+                data_name, data_type = self._deserialize_shared_tensor(data_str)
+                dtype = F.data_type_dict[data_type]
+                data_shape = F.asnumpy(msg.id).tolist()
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py:482]: this looks weird. why is `data_shape` generated from `msg.id`?

File: tests/compute/test_kvstore.py [hunk 5]
```diff
@@ -40,10 +40,13 @@ def start_client():
     my_client = KVClient(server_namebook=server_namebook)
     my_client.connect()
 
+    my_client.init_data(name='data_2', shape=[num_entries, dim_size], dtype=F.float32, target_name='data_0')
```
Comment (zheng-da) [inline tests/compute/test_kvstore.py:43]: so target name is some existing tensor that uses the same ID space.

**Comment by aksnzhy** [comment]
This PR is ready to merge. @zheng-da

File: python/dgl/contrib/dis_kvstore.py (hunk 1 above)
Comment (aksnzhy) [reply python/dgl/contrib/dis_kvstore.py:481]: Yes. And other clients will just recv the message.

File: python/dgl/contrib/dis_kvstore.py (hunk 4 above)
Comment (aksnzhy) [reply python/dgl/contrib/dis_kvstore.py]: Changed.

File: src/graph/network.cc [hunk 6]
```diff
@@ -499,14 +507,19 @@ static void send_kv_message(network::Sender* sender,
...
-    send_id_msg.deallocator = [id](Message*) {};
-    CHECK_EQ(sender->Send(send_id_msg, recv_id), ADD_SUCCESS);
+    if (kv_msg->msg_type != kInitMsg) {
+      Message send_id_msg;
+      send_id_msg.data = static_cast<char*>(kv_msg->id->data);
+      send_id_msg.size = kv_msg->id.GetSize();
+      NDArray id = kv_msg->id;
+      if (auto_free) {
+        send_id_msg.deallocator = [id](Message*) {};
+      }
```
Comment (zheng-da) [inline src/graph/network.cc:517]: you add `auto_free` check here. I suppose this is a bug fix?

File: src/graph/network.cc [hunk 7]
```diff
@@ -516,6 +529,19 @@ static void send_kv_message(network::Sender* sender,
...
+        kv_msg->msg_type != kPullBackMsg) {
+      Message send_shape_msg;
+      send_shape_msg.data = static_cast<char*>(kv_msg->shape->data);
+      send_shape_msg.size = kv_msg->shape.GetSize();
+      NDArray shape = kv_msg->shape;
+      if (auto_free) {
+        send_shape_msg.deallocator = [shape](Message*) {};
+      }
+      CHECK_EQ(sender->Send(send_shape_msg, recv_id), ADD_SUCCESS);
+    }
```
Comment (zheng-da) [inline src/graph/network.cc:544]: do you send each field `Id, shape, data` in separate messages?

**Pull request merged**